Coupons: Add Limit section missing fields to Coupon Details

ThomazFB opened the pull request:
Summary
==========
Fix issue #6693 by introducing two missing usage limits fields into the Coupon Details view: Usage limit per coupon and Usage limit by item amount in the cart.

Screenshots
==========
| Limit info in plural  | Limit info in singular |
| ------------- | ------------- |
| ![Simulator Screen Shot - iPhone 13 Pro Max - 2022-05-11 at 23 07 04](https://user-images.githubusercontent.com/5920403/167977828-4fc224bb-807a-4bd1-b532-bb44ad471d3e.png) | ![Simulator Screen Shot - iPhone 13 Pro Max - 2022-05-11 at 23 06 59](https://user-images.githubusercontent.com/5920403/167977836-3b438d6d-1c9c-4e6d-990b-c49b78def1d3.png) |

| Android Coupon details view  | iOS coupon details view |
| ------------- | ------------- |
| ![Screenshot_20220511_231531](https://user-images.githubusercontent.com/5920403/167978661-8f9f273c-3ca9-4286-bf9a-8cd9ec90c75a.png) | ![Simulator Screen Shot - iPhone 13 Pro Max - 2022-05-11 at 23 07 04](https://user-images.githubusercontent.com/5920403/167977828-4fc224bb-807a-4bd1-b532-bb44ad471d3e.png) |




How to Test
==========
1. Make sure the app has the Coupons experimental features toggle activated
2. Make sure there are at least two coupons configured with the Usage limits configuration, one with the `Usage limit per coupon` and `Limit usage to X Items` fields set to any number higher than 1, the other with both fields set as exactly 1, and a third coupon with none of these fields set.
3. Go to the Coupon list and select these three configured Coupons, the one with the limits set as higher than 1 must display data using plural info, the one with the limits set as 1 must display data using singular info, and the one with no limits set must not display anything related to those fields.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Define itemLimitUsage and usageLimitPerCoupon string resources

ThomazFB committed:
Define usageLimit and limitUsageToXItems in CouponDetailsViewModel

ThomazFB committed:
populate usageLimit and limitUsageToXItems in CouponDetailsViewModel with coupon model data

ThomazFB committed:
Add usageLimit and limitUsageToXItems info to CouponDetails view

ThomazFB committed:
Fix CouponDetails new strings constraints and positioning

ThomazFB committed:
Refactor itemsInCart and usageLimitPerCoupon string resources names and comment

ThomazFB committed:
Split itemsInCart and usageLimitPerCoupon string resources between singular and plural versions

ThomazFB committed:
Add singular and plural versions of itemsInCart and usageLimitPerCoupon string resources with respective constraints

ThomazFB committed:
Fix code spacing in CouponDetails new changes

ThomazFB committed:
Adjust CouponDetailsViewModelTests to include usageLimit and limitUsageToXItems

ThomazFB committed:
Reformat code spacing in limit properties

ThomazFB commented on WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetails.swift:
```diff
@@ -256,11 +256,23 @@ struct CouponDetails: View {
                 Text(String.localizedStringWithFormat(Localization.maximumSpend, viewModel.maximumAmount))
                     .renderedIf(viewModel.maximumAmount.isNotEmpty)
 
+                Text(String.localizedStringWithFormat(Localization.singularUsageLimitPerCoupon, viewModel.usageLimit))
+                    .renderedIf(viewModel.usageLimit == 1)
+
+                Text(String.localizedStringWithFormat(Localization.pluralUsageLimitPerCoupon, viewModel.usageLimit))
+                    .renderedIf(viewModel.usageLimit > 1)
+
```
I'm not sure if the limits info I added here should be within the rendering constraints of this VStack, but I decided to follow the info order used in the Android implementation as a reference, also considering that the other limit info that was already here is also inserted inside this VStack. Let me know if this doesn't make sense.

itsmeichigo replied:
The reason why I added the condition check to the outer VStack was that the VStack was defined with spacing, so if none of the details in this VStack is rendered, we still get a small space like so:

<img src="https://user-images.githubusercontent.com/5533851/167989704-cb3e144a-4a83-46e7-93fd-1b2b396dda71.png" width=320 />

So, to be on the safe side, we need the conditions on both the views and the containing VStacks 😅 

An alternative solution would be to add padding to each of the items and remove the spacing for the VStack, but we'll end up with some duplicated code too. There's no perfect solution here to be honest.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetails.swift:
```diff
@@ -256,11 +256,23 @@ struct CouponDetails: View {
                 Text(String.localizedStringWithFormat(Localization.maximumSpend, viewModel.maximumAmount))
                     .renderedIf(viewModel.maximumAmount.isNotEmpty)
 
+                Text(String.localizedStringWithFormat(Localization.singularUsageLimitPerCoupon, viewModel.usageLimit))
+                    .renderedIf(viewModel.usageLimit == 1)
+
+                Text(String.localizedStringWithFormat(Localization.pluralUsageLimitPerCoupon, viewModel.usageLimit))
+                    .renderedIf(viewModel.usageLimit > 1)
+
```
Suggestion: we can simplify this using the `pluralize` extension method for `String`. The only problem is that we're supporting `Int` amount only, should we add a new method to support `Int64`? With that we can make it simpler like so:

```suggestion
                Text(String.pluralize(viewModel.usageLimit,
                                      singular: Localization.singularUsageLimitPerCoupon,
                                      plural: Localization.pluralUsageLimitPerCoupon))
                    .renderedIf(viewModel.usageLimit > 0)
```

WDYT?

ThomazFB replied:
Good one! I've fixed it in both these commits: https://github.com/woocommerce/woocommerce-ios/pull/6831/commits/1e0a4060d199e0ece9bf9353347bdd4b02ff485d and https://github.com/woocommerce/woocommerce-ios/pull/6831/commits/a312d0fb86d7fb9ccdfed2bb78210c557e908766.

But I have one newbie question: Do we still support 32bit devices in our iOS versions? If not, wouldn't every `Int` already be by default an `Int64`? Or Swift forces every `Int` to be `Int32`? I ask this because I was thinking if these properties shouldn't be converted directly to `Int`, this way we could drop the pluralize I added for `Int64` and only use the existing one. 🤔

itsmeichigo replied:
Great question! 💯 It is true that we can actually cast `Int64` to `Int` since Apple dropped support for 32bit systems and forced all apps to support 64bit since iOS 11. I think we're still using `Int64` to explicitly state the support for IDs and values that can be quite large. This also keeps it consistent with the practice we've been doing since the "old days".

With that said, I believe can drop the `Int64` variant of the `pluralize` method - thank you for pointing that out!

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetails.swift:
```diff
@@ -256,11 +256,23 @@ struct CouponDetails: View {
                 Text(String.localizedStringWithFormat(Localization.pluralLimitPerUser, viewModel.usageLimitPerUser))
                     .renderedIf(viewModel.usageLimitPerUser > 1)
+
+                Text(String.localizedStringWithFormat(Localization.singularItemsInCartUsageLimit, viewModel.limitUsageToXItems))
+                    .renderedIf(viewModel.limitUsageToXItems == 1)
+
+                Text(String.localizedStringWithFormat(Localization.pluralItemsInCartUsageLimit, viewModel.limitUsageToXItems))
+                    .renderedIf(viewModel.limitUsageToXItems > 1)
```
We can also simplify this following the suggestion for `usageLimit` above.

ThomazFB replied:
Sure! Done in https://github.com/woocommerce/woocommerce-ios/commit/1e0a4060d199e0ece9bf9353347bdd4b02ff485d and https://github.com/woocommerce/woocommerce-ios/commit/a312d0fb86d7fb9ccdfed2bb78210c557e908766.

I also took the liberty and adapted the `usageLimitPerUser` to also take advantage of the new pluralization function.

itsmeichigo replied:
Great, thanks!

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetails.swift:
```diff
@@ -379,6 +391,26 @@ private extension CouponDetails {
+        static let singularItemsInCartUsageLimit = NSLocalizedString(
+                "Limited to %1$d item in cart",
+                comment: "The singular limit of the same item in cart to apply a coupon, reads like: " +
+                        "Limited to 10 item in cart"
```
Nit: should this be 1 item instead? 😆 

```suggestion
                        "Limited to 1 item in cart"
```

ThomazFB replied:
Oops 😬, thanks! Done here: https://github.com/woocommerce/woocommerce-ios/pull/6831/commits/65e6913a5a970f5a1edbc6b2ab77a7b0e15e19a5

itsmeichigo reviewed (commented):
Great work Thomaz! Regarding the condition checks in the outer `VStack` of these new views, I think we still need to check that `usageLimit` and `limitUsageToXItems` are larger than 0. I added the explanation and more suggestions below.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetails.swift:
```diff
@@ -379,6 +391,26 @@ private extension CouponDetails {
         )
+        static let singularItemsInCartUsageLimit = NSLocalizedString(
+                "Limited to %1$d item in cart",
+                comment: "The singular limit of the same item in cart to apply a coupon, reads like: " +
```
Nit: maybe we can reword it to make it sound clearer to translators:

```suggestion
                comment: "The required number of items in the cart to apply a coupon in singular form, reads like: " +
```

ThomazFB replied:
done here: https://github.com/woocommerce/woocommerce-ios/pull/6831/commits/65e6913a5a970f5a1edbc6b2ab77a7b0e15e19a5

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Coupons/CouponDetails/CouponDetails.swift:
```diff
@@ -379,6 +391,26 @@ private extension CouponDetails {
+        )
+        static let pluralItemsInCartUsageLimit = NSLocalizedString(
+                "Limited to %1$d items in cart",
+                comment: "The plural limit of the same item in cart to apply a coupon, reads like: " +
```
Another suggestion:
```suggestion
                comment: "The required number of items in the cart to apply a coupon in plural form, reads like: " +
```

ThomazFB replied:
Nice! done here: https://github.com/woocommerce/woocommerce-ios/pull/6831/commits/65e6913a5a970f5a1edbc6b2ab77a7b0e15e19a5

ThomazFB committed:
Refactor CouponDetails limit strings

ThomazFB committed:
Add Int64 pluralize support in String+Helpers

ThomazFB committed:
Adjust limit fields to take advantage of pluralize function

ThomazFB committed:
Adjust outer VStack to also consider new limits fields rendering constraints

ThomazFB committed:
Fix trailing whitespace violations

ThomazFB commented:
Thank you for your amazing review and tips @itsmeichigo 🙇! I've addressed all your suggestions and the PR is ready for a second round!

itsmeichigo reviewed (approved):
Thanks for the updates Thomaz!

ThomazFB committed:
Remove Int64 pluralize version

ThomazFB commented:
Thanks @itsmeichigo for approving the PR. As follow up on our discussion, I removed the pluralize with the `Int64` parameter I've added previously, keeping the properties as `Int64` to the project coding standard, and adopted the `Int(truncatingIfNeeded: BinaryInteger)` initializer to cast the properties inside the SwiftUI file, WDYT?

Changes done here: https://github.com/woocommerce/woocommerce-ios/pull/6831/commits/73f412fe9ea8ff034c13dc2f691677f65690f8e8

itsmeichigo commented:
> adopted the Int(truncatingIfNeeded: BinaryInteger) initializer to cast the properties inside the SwiftUI file

I'm thinking using the `truncatingIfNeeded` constructor is a bit safer than necessary since as you know, `Int` is the same size as `Int64` on 64bit platforms and we can probably get away with using the simple constructor `Int(<Int64>)`. But it's ok to be extra safe too, so I'm good either way.

ThomazFB committed:
Replace Int initializer with truncating operation for the standard Int64 to Int initializer
